### PR TITLE
Namenode ha plugin

### PR DESCRIPTION
**Pull request by agardella**
@captain695 @cburk

### CODE REVIEW

**Comment by pelmers** [comment]
can you remove the .DS_STORE files?

File: rice-namenode/source/main.cc [hunk 1]
```diff
@@ -29,5 +30,8 @@ int main(int argc, char* argv[]) {
 	nncli.register_watches();
 	std::cout << "Namenode is starting" << std::endl;
 	ClientNamenodeTranslator translator(port, nncli);
-	translator.getRPCServer().serve(io_service);
```
Comment (nichhk) [inline rice-namenode/source/main.cc:35]: Why are we able to delete this?

File: rice-namenode/source/HAServiceProtocolImpl.cc [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+
+// ------------------------------- HELPERS -----------------------------
+
+void HaServiceTranslator::logMessage(google::protobuf::Message& req, std::string req_name) {
+	LOG(INFO) << CLASS_NAME <<  "Got message " << req_name << ": " << req.DebugString();
+}
+
+HaServiceTranslator::~HaServiceTranslator() {
```
Comment (nichhk) [inline rice-namenode/source/HAServiceProtocolImpl.cc]: Destructor is not necessary.

**Review by nichhk — CHANGES_REQUESTED** [review]
We can have everything start in the ACTIVE state. And transitionToActive and transitionToStandby don't need to do anything.

File: vagrant_provision.sh [hunk 3]
```diff
@@ -52,9 +52,21 @@ cat > /home/vagrant/hadoop3/etc/hadoop/hdfs-site.xml <<EOF
     </property>
 </configuration>
 EOF
+
+# add custom config files to Hadoop
+cat /home/vagrant/rdfs/config/hdfs-site.xml > /home/vagrant/hadoop/etc/hadoop/hdfs-site.xml
+cat /home/vagrant/rdfs/config/core-site.xml > /home/vagrant/hadoop/etc/hadoop/core-site.xml
```
Comment (pelmers) [inline vagrant_provision.sh:39]: can you remove the lines above this which touch the same files but get overwritten here?
`cat > /home/vagrant/hadoop3/etc/hadoop/core-site.xml <<EOF`
and
`cat > /home/vagrant/hadoop3/etc/hadoop/hdfs-site.xml <<EOF`

File: vagrant_provision.sh [hunk 4]
```diff
@@ -52,9 +52,21 @@ cat > /home/vagrant/hadoop3/etc/hadoop/hdfs-site.xml <<EOF
...
+
 # add hadoop to path
 echo 'export PATH=/home/vagrant/hadoop/bin:$PATH' >> /home/vagrant/.bashrc
 
+# add hadoop to classpath
+echo 'export CLASSPATH=/home/vagrant/hadoop/share/hadoop/hdfs/*:/home/vagrant/hadoop/share/hadoop/common/*' >> /home/vagrant/.bashrc
+
+# add diff detector to path
+echo 'python /home/vagrant/rdfs/utility/provision_diff.py' >> /home/vagrant/.bashrc
```
Comment (pelmers) [inline vagrant_provision.sh]: this line is repeated at the end of the file too, delete this one

**Comment by Vegetable26** [comment]
Have we tried the scenario where we have two NameNodes A and B. Kill A, then kill B, then bring back A?

File: rice-namenode/include/HaServiceProtocolImpl.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		 */
+		void logMessage(google::protobuf::Message& req, std::string req_name);
+
+		/**
+		 * Get an rpc header proto given an error message and exception classname
+		 */
+		hadoop::common::RpcResponseHeaderProto GetErrorRPCHeader(std::string error_msg,
+        		std::string exception_classname);
+
+        static const int LEASE_CHECK_TIME; 	// in seconds, how often the namenode checks all leases
```
Comment (efrederic) [inline rice-namenode/include/HaServiceProtocolImpl.h]: is LEASE_CHECK_TIME ever used?

Comment (nichhk) [reply rice-namenode/include/HaServiceProtocolImpl.h]: Nope leases are gone.

File: rice-namenode/source/main.cc (hunk 1 above)
Comment (agardella) [reply rice-namenode/source/main.cc:35]: Lines 34 and 36 do the same thing

File: vagrant_provision.sh (hunk 3 above)
Comment (agardella) [reply vagrant_provision.sh:39]: I think they're touching different files (the ones in /home/vagrant/hadoop3 as opposed to /home/vagrant/hadoop). It looks like they might be unnecessary anyway, but I'm not totally sure

File: vagrant_provision.sh (hunk 4 above)
Comment (agardella) [reply vagrant_provision.sh]: fixed

File: rice-namenode/include/HaServiceProtocolImpl.h (hunk 5 above)
Comment (agardella) [reply rice-namenode/include/HaServiceProtocolImpl.h]: fixed

File: rice-namenode/source/HAServiceProtocolImpl.cc (hunk 2 above)
Comment (agardella) [reply rice-namenode/source/HAServiceProtocolImpl.cc]: fixed

File: vagrant_provision.sh (hunk 3 above)
Comment (pelmers) [reply vagrant_provision.sh:39]: they are the same file, `hadoop` is just a link to `hadoop3` because of `ln -s hadoop3 hadoop` a few lines above.

**Comment by captain695** [comment]
@Vegetable26 just tried that scenario and it seems to work

**Comment by captain695** [comment]
@nichhk does this look all good to merge?

**Comment by nichhk** [comment]
@captain695 LGTM

**Comment by captain695** [comment]
@nichhk I know you guys have been having things break in develop. Should I wait until that's resolved until I try to merge this in?

**Comment by agardella** [comment]
@nichhk Having all of the namenodes start in active means we can't do a manual failover, and it changes quite a bit of our presentation and demo... it also seems conceptually wrong. What's the reason you wanted that change?

**Pull request merged**